File: plugins/collector/event/configs/sshconfig.go
package configs

import (
	"bufio"
	"collector/cache/user"
	"collector/eventmanager"
	"collector/utils"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/bytedance/sonic"
	"github.com/chriskaliX/SDK"
	"github.com/chriskaliX/SDK/transport/protocol"
	"github.com/mitchellh/mapstructure"
)

const (
	userSshConfig      = ".ssh/config"
	systemSshConfig    = "/etc/ssh/ssh_config"
	SSHCONFIG_DATATYPE = 3005
)

var _ eventmanager.IEvent = (*SshConfig)(nil)

type SshConfig struct {
	// check if first time, pay attention to set true
	firstTime bool
}

func (SshConfig) DataType() int { return SSHCONFIG_DATATYPE }

func (SshConfig) Name() string { return "sshconfig" }

func (n *SshConfig) Flag() eventmanager.EventMode { return eventmanager.Periodic }

func (SshConfig) Immediately() bool { return false }

func (s *SshConfig) Run(sandbox SDK.ISandbox, sig chan struct{}) error {
	hash := utils.Hash()
	// get user configuration
	configPath := s.sshConfigPath()
	s.firstTime = true
	configs := make([]sshConfig, 0, 20)
	for uid, path := range configPath {
		if config, err := s.getSshConfig(uid, path); err == nil {
			configs = append(configs, config...)
		}
	}
	// get system configuration
	if config, err := s.getSshConfig("0", systemSshConfig); err == nil {
		configs = append(configs, config...)
	}

	for _, config := range configs {
		rec := &protocol.Record{
			DataType: int32(s.DataType()),
			Data: &protocol.Payload{
				Fields: make(map[string]string, 5),
			},
		}
		mapstructure.Decode(&config, &rec.Data.Fields)
		rec.Data.Fields["package_seq"] = hash
		sandbox.SendRecord(rec)
	}
	return nil
}

// Depend on usercache, execute after GetUser
func (SshConfig) sshConfigPath() (configs map[string]string) {
	configs = make(map[string]string)
	users := user.Cache.GetUsers()
	for _, user := range users {
		configs[user.UID] = filepath.Join(user.HomeDir, userSshConfig)
	}
	return
}

type sshConfig struct {
	Uid      string `json:"uid" mapstructure:"uid"`
	Block    string `json:"block" mapstructure:"block"`
	Option   string `mapstructure:"option"`
	Filepath string `json:"filepath" mapstructure:"filepath"`
	option   map[string]string
}

// Reference:
// https://github.com/osquery/osquery/blob/d2be385d71f401c85872f00d479df8f499164c5a/osquery/tables/system/ssh_configs.cpp
func (s *SshConfig) getSshConfig(uid string, path string) (configs []sshConfig, err error) {
	var (
		file   *os.File
		scan   *bufio.Scanner
		config = sshConfig{
			option: make(map[string]string),
		}
	)
	if file, err = os.Open(path); err != nil {
		return
	}
	defer file.Close()
	scan = bufio.NewScanner(io.LimitReader(file, 16834))
	for scan.Scan() {
		text := strings.TrimSpace(scan.Text())
		text = strings.ToLower(text)
		if len(text) == 0 || text[0] == '#' {
			continue
		}
		// if matches the host or match fields, start record configuration
		if strings.HasPrefix(text, "host ") || strings.HasPrefix(text, "match ") {
			if s.firstTime {
				s.firstTime = false
			} else {
				configs = append(configs, config)
			}
			config = sshConfig{
				option: make(map[string]string),
			}
			config.Block = text
			config.Filepath = path
			config.Uid = uid
		} else {
			// don't know it's ` ` or `=`, try everytime
			spaceIndex := strings.Index(text, " ")
			equalIndex := strings.Index(text, "=")
			if spaceIndex == -1 && equalIndex == -1 {
				config.option[text] = ""
			} else if spaceIndex == -1 {
				config.option[text[:equalIndex]] = text[equalIndex+1:]
			} else {
				config.option[text[:spaceIndex]] = text[spaceIndex+1:]
			}
		}
	}
	config.Option, _ = sonic.MarshalString(config.option)
	configs = append(configs, config)
	return
}

func init() { addEvent(&SshConfig{}) }

File: plugins/collector/src/event/configs/sshconfig.rs
/// SshConfig collector — data_type 3005.
/// Parses /etc/ssh/ssh_config + per-user ~/.ssh/config.
/// Mirrors Go's event/configs/sshconfig.go.
use std::collections::HashMap;
use std::fs;
use std::io::{BufRead, BufReader, Read};

use anyhow::Result;
use sdk::Client;

use crate::event::{hash, make_record};

const DATA_TYPE: i32 = 3005;
const SYSTEM_SSH_CONFIG: &str = "/etc/ssh/ssh_config";
const USER_SSH_CONFIG:   &str = ".ssh/config";

/// One SSH config block (mirrors Go's sshConfig struct).
struct SshConfigBlock {
    uid:      String,
    block:    String,  // "Host xxx" or "Match xxx"
    option:   String,  // JSON of key→value map
    filepath: String,
}

pub async fn run(client: &mut Client) -> Result<()> {
    let hash_val = hash();
    let mut configs: Vec<SshConfigBlock> = Vec::new();

    // Per-user configs (read /etc/passwd for home dirs)
    for (uid, home) in user_homes() {
        let path = format!("{}/{}", home.trim_end_matches('/'), USER_SSH_CONFIG);
        if let Ok(c) = parse_ssh_config(&uid, &path) {
            configs.extend(c);
        }
    }
    // System config (uid "0")
    if let Ok(c) = parse_ssh_config("0", SYSTEM_SSH_CONFIG) {
        configs.extend(c);
    }

    for cfg in configs {
        let mut fields = HashMap::new();
        fields.insert("uid".into(),         cfg.uid);
        fields.insert("block".into(),       cfg.block);
        fields.insert("option".into(),      cfg.option);
        fields.insert("filepath".into(),    cfg.filepath);
        fields.insert("package_seq".into(), hash_val.clone());
        let _ = client.send_record(&make_record(DATA_TYPE, fields));
    }
    Ok(())
}

/// Parse an SSH config file into blocks (Host/Match blocks → options JSON).
fn parse_ssh_config(uid: &str, path: &str) -> Result<Vec<SshConfigBlock>> {
    let f = fs::File::open(path)?;
    let reader = BufReader::new(f.take(16 * 1024));

    let mut blocks: Vec<SshConfigBlock> = Vec::new();
    let mut current_block = String::new();
    let mut current_opts: HashMap<String, String> = HashMap::new();
    let mut first = true;

    let flush = |block: &str, opts: &HashMap<String, String>, filepath: &str, uid: &str, out: &mut Vec<SshConfigBlock>| {
        if block.is_empty() { return; }
        let option = serde_json::to_string(opts).unwrap_or_default();
        out.push(SshConfigBlock {
            uid:      uid.to_owned(),
            block:    block.to_owned(),
            option,
            filepath: filepath.to_owned(),
        });
    };

    for line in reader.lines().map_while(Result::ok) {
        let text = line.trim().to_lowercase();
        if text.is_empty() || text.starts_with('#') { continue; }

        if text.starts_with("host ") || text.starts_with("match ") {
            if first { first = false; }
            else { flush(&current_block, &current_opts, path, uid, &mut blocks); }
            current_block = text.clone();
            current_opts.clear();
        } else {
            // Go priority: space index first, then equals — mirrors ssh_configs.cpp
            let space_idx  = text.find(' ');
            let equals_idx = text.find('=');
            let (k, v) = match (space_idx, equals_idx) {
                (None, None)       => (text.as_str(), ""),
                (None, Some(ei))   => (text[..ei].trim(),  text[ei+1..].trim()),
                (Some(si), None)   => (text[..si].trim(),  text[si+1..].trim()),
                (Some(si), Some(_))=> (text[..si].trim(),  text[si+1..].trim()),
            };
            current_opts.insert(k.to_owned(), v.to_owned());
        }
    }
    flush(&current_block, &current_opts, path, uid, &mut blocks);
    Ok(blocks)
}

/// Read /etc/passwd to get (uid_string, home_dir) pairs.
fn user_homes() -> Vec<(String, String)> {
    let Ok(f) = fs::File::open("/etc/passwd") else { return Vec::new() };
    BufReader::new(f)
        .lines()
        .map_while(Result::ok)
        .filter_map(|line| {
            let parts: Vec<&str> = line.splitn(7, ':').collect();
            if parts.len() >= 6 {
                Some((parts[2].to_owned(), parts[5].to_owned()))
            } else { None }
        })
        .collect()
}
